### ISSUE TITLE
Static Optimization fails with ligaments

### ISSUE THREAD

**Issue by jimmyDunne** [opensim-org/opensim-core#340]
Lianne Zevenbergen reported an issue with Static optimization

>In the current version of OpenSim 3.2 it is not possible to perform a static >optimization when an object is included of the ligament class. We have found >the bug in the Source code of OpenSim3.2 to solve this problem.
 
>The ligament class is part of the ForceSet in an OpenSim model. The ligament >force is determined by its length which is defined by its path (path points + >coordinates of the joint) and therefore doesn’t fall under the actuator class. In >static optimization the force produced by the actuators is determined by >minimizing the sum of squared actuator activations to resolve the net joint >moments as determined by the inverse dynamics.
 
>In the source file StaticOptimizationTarget.ccp the function computeAcceleration >is defined. In this function the j-index, to update the parameters and optimal >force, is updated incorrectly. Instead of increasing j only if the force is an >actuator, it is now updated for every force. This gives problems when not all the >forces in the model are actuators, like with ligaments. We suggest to update the >source code as follows:

``` C++
       const ForceSet& fs = _model->getForceSet();
       for(int i=0,j=0;i<fs.getSize();i++)  {
         Actuator *act = dynamic_cast<Actuator*>(&fs.get(i));
             if( act ) {
             act->setOverrideForce(s,parameters[j]*_optimalForce[j]);
                    j++; // New position to update j
             }
         // j++ (old position)
    }
```
>In this case the constraintMatrix is updated correctly to solve the static >optimization problem.
 
>Another suggestion is to change the function constructCollumnLabels in >StaticOptimization.cpp  in such way that only the actuator forces are included in >the header of the output files.
``` C++
void StaticOptimizationLig::
constructColumnLabels()
{
       Array<string> labels;
       labels.append("time");
       if(_model)
             for (int i=0; i < _forceSet->getSize(); i++) {
                    Actuator* act = dynamic_cast<Actuator*>(&_forceSet->get(i));
                    if( act )labels.append(act->getName());
             }
       // old
       // if(_model)
       //     for (int i=0; i < _forceSet->getSize(); i++) labels.append(_forceSet->get(i).getName());
       setColumnLabels(labels);
}
```
> In this way the header labels correspond with the output generated by the    
> optimization results (forces and activations).
 
>We have implemented these changes and now static optimization works as >expected.

**Comment by chrisdembia** [comment]
Awesome! Should we ask them to submit a pull request?

**Comment by jimmyDunne** [comment]
This will be a bigger overhaul of SO and tools

**Comment by jimmyDunne** [comment]
SO solver should take possession of the muscle actuators and other forces should just be left alone. @aseth1 will clarify

**Comment by aymanhab** [comment]
Addressed by #546
